server/server.go

```diff
@@ -3652,7 +3791,12 @@ func (s *StanServer) processSubscriptionRequest(m *nats.Msg) {
 
 		// Case of restarted durable subscriber, or first durable queue
 		// subscriber re-joining a group that was left with pending messages.
-		err = s.updateDurable(ss, sub)
+		if s.isClustered() {
+			// If clustered, thread update through Raft.
+			sub, err = s.replicateUpdateSubscription(c, &sub.SubState, sr.Subject)
```

Comment:
If I understand correctly, here we are the leader and replicate the subscription create/update protocol. The followers should receive that and store record on disk. For a follower, there is no need for the `sub` object since it doesn't do anything with it. For the leader, it has created `sub` already above and now will have `sub` recreated in the replicate code, except that it won't have all the things set/reset that has been done above. So why create `sub` (and return it as a future) in `Apply()`? This seem broken to me. For the leader, it would make more sense to pass in the replication protocol some key that could be used to retrieve the sub that is already created, or move all the code reseting the durable subscription into Apply(), but even then, it is problematic since `sub` was stored in memory and you are now replacing it with something completely different.

Reply:
Not sure if I completely follow. The actual creation needs to happen in `Apply` since that is guaranteed to execute cluster-wide. I agree it's a bit weird we create the sub struct twice basically, but I'm not sure how to do this otherwise since that struct isn't serializable.

Reply:
Look at line 3754 in your branch. This is the case where this is a remembered subscription, that is, a durable one. `sub` is already created and stored in some queue state or `ss.durables` for non queue durable subscriptions. Some setting on the `sub` is done when the durable restarts. But your code will have 2 effects: override `sub` with whatever is created in `Apply()` which won't contain the same reset fields than done above and also will not be stored in the durables map..
If still not clear, let's discuss this.

Reply:
I think I see what you're saying but IMO the `processSubscriptionRequest` code is really complex and hard for me to follow what all is going on. I do know that not returning the sub from `Apply` caused this to not work (don't remember specifics but that was an issue). We should probably discuss this in more detail offline.